src/main/java/hudson/tasks/junit/CaseResult.java

```diff
@@ -256,10 +262,25 @@ private static String getSkippedMessage(Element testCase) {
         return message;
     }
 
-    public String getDisplayName() {
+    public String getTransformedTestName() {
         return TestNameTransformer.getTransformedName(testName);
     }
 
+    public String getDisplayName() {
+        // Only prepend the enclosing flow node names if there are any and the run this is in has multiple blocks directly containing
+        // test results.
+        if (!getEnclosingFlowNodeNames().isEmpty()) {
```

Comment:
I agree it's *extremely* useful for the user, I'm just deeply concerned about the performance implications of repeatedly scanning the flowgraph from end to start -- for example, in the common case of parallel test runs, we'd have N branches with M nodes preceding this, and archiving JUnit results would be O(m*n) overall, which would probably be surprising to users. 

Scanning for the nearest enclosing stage can *potentially* exhibit that performance, but in practice that is very where.  Could we extend the cache from https://github.com/jenkinsci/workflow-api-plugin/pull/45 to provide the unclosed blocks?